crates/bevy_ecs/src/schedule/run_criteria.rs

```diff
@@ -0,0 +1,244 @@
+use crate::{
+    archetype::{Archetype, ArchetypeComponentId},
+    component::ComponentId,
+    query::Access,
+    schedule::{BoxedRunCriterionLabel, RunCriterionLabel},
+    system::{BoxedSystem, System, SystemId},
+    world::World,
+};
+use std::borrow::Cow;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ShouldRun {
+    /// Yes, the system should run.
+    Yes,
+    /// No, the system should not run.
+    No,
+    /// Yes, the system should run, and afterwards the criteria should be checked again.
+    YesAndCheckAgain,
+    /// No, the system should not run right now, but the criteria should be checked again later.
+    NoAndCheckAgain,
+}
+
+pub(crate) struct RunCriterion {
```

Comment:
I think I still prefer using "run criteria" instead of "run criterion" everywhere:

1. If I was phrasing this in normal english, I would ask "what criteria do you want to use to run this system?". A given "run criteria system" can use any number of criteria to determine whether or not the system ShouldRun.
2. Criterion is basically never used in conversational english and therefore feels "unnatural" in this context.
3. Criteria is shorter

Reply:
Tentatively, I agree.

I'd also like to float `Condition` as an alternative.

Reply:
Condition is much less arcane and makes the connection to conditional execution more obvious.

Reply:
"Condition" works for me.

Reply:
I've renamed everything back to "run criteria" for now. We can bikeshed/change it in a different PR.